File: code_examples/python_bindings/python/algorithms.py
def find_primes(limit):
    """Sieve of Eratosthenes in pure Python"""
    is_prime = [True] * (limit + 1)
    is_prime[0] = is_prime[1] = False
    
    for i in range(2, int(limit**0.5) + 1):
        if is_prime[i]:
            for j in range(i*i, limit + 1, i):
                is_prime[j] = False
    
    return [i for i in range(2, limit + 1) if is_prime[i]]

def matrix_multiply(a, b, n):
    """Matrix multiplication in pure Python"""
    c = [0.0] * (n * n)
    
    for i in range(n):
        for j in range(n):
            sum_val = 0.0
            for k in range(n):
                sum_val += a[i * n + k] * b[k * n + j]
            c[i * n + j] = sum_val
    
    return c

def fibonacci(n):
    """Fibonacci with memoization in pure Python"""
    if n <= 1:
        return n
    return fibonacci(n - 1) + fibonacci(n - 2)

File: code_examples/python_bindings/src/algorithms.cpp
#include "algorithms.h"
#include <cmath>
#include <random>
#include <unordered_map>
#include <algorithm>

std::vector<int> find_primes(int limit) {
    std::vector<bool> is_prime(limit + 1, true);
    std::vector<int> primes;
    
    is_prime[0] = is_prime[1] = false;
    
    for (int i = 2; i * i <= limit; ++i) {
        if (is_prime[i]) {
            for (int j = i * i; j <= limit; j += i) {
                is_prime[j] = false;
            }
        }
    }
    
    for (int i = 2; i <= limit; ++i) {
        if (is_prime[i]) {
            primes.push_back(i);
        }
    }
    
    return primes;
}

std::vector<double> matrix_multiply(const std::vector<double>& a,
                                   const std::vector<double>& b,
                                   int n) {
    std::vector<double> c(n * n, 0.0);
    
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            double sum = 0.0;
            for (int k = 0; k < n; ++k) {
                sum += a[i * n + k] * b[k * n + j];
            }
            c[i * n + j] = sum;
        }
    }
    
    return c;
}

long long fibonacci(int n) {
    if (n <= 1) return n;
    return fibonacci(n - 1) + fibonacci(n - 2);
}
